### PR TITLE
feat(grpc_datasource): add ConnectRPC transport implementation

### PR DESCRIPTION
**Pull request by fengyuwusong**
## Summary

Add a ConnectRPC implementation of the `RPCTransport` interface introduced in #1490, so the gRPC data source can dial Connect subgraphs without changing the compiler or JSON builder. The same `MockService` fixture is regenerated via `buf` to emit Connect handlers alongside the existing gRPC code, and a new end-to-end test exercises the data source over Connect against that fixture.

This is the second of the two PRs that together replace #1480.

## Related

- Builds on: #1490 (`RPCTransport` interface + `GRPCTransport` implementation)
- Replaces: #1480 (the original combined PR, now split into #1490 + this one)
- Consumed by: wundergraph/cosmo#2808 (the router-side feature that motivated this work)

### `pkg/grpctest` — buf-driven fixture

- New `buf.yaml` / `buf.gen.yaml`: drive `protoc-gen-go`, `protoc-gen-go-grpc`, and `protoc-gen-connect-go` from a single `buf generate` invocation, keeping the existing import path (`graphql-go-tools/v2/pkg/grpctest/productv1`) via `M=` overrides.
- `Makefile`: `generate-proto` now calls `buf generate`; `install-protoc` also installs `protoc-gen-connect-go`.
- New `productv1/productv1connect/product.connect.go`: buf-generated Connect handler/client. Sibling of the existing `productv1/product_grpc.pb.go`, so the package serves Connect, gRPC, and gRPC-Web from one HTTP endpoint.
- New `mockservice_connect.go`: hand-maintained adapter (`MockServiceConnect`) wrapping the existing `MockService` onto `productv1connect.ProductServiceHandler`. Same backing implementation, two transports.

### `pkg/engine/datasource/grpc_datasource` — ConnectRPC transport

- New `transport_connect.go`: `NewConnectTransport(ConnectTransportConfig{BaseURL, Encoding, Interceptors, HTTPClient})` implements `RPCTransport.Invoke` on top of `connectrpc.com/connect`. Supports `ConnectEncodingProtobuf` ("proto") and `ConnectEncodingJSON` ("json") wire formats. Translates `grpc/metadata` to and from Connect headers using the same conventions as `GRPCTransport`, with binary request headers (`-bin` suffix) base64-encoded per the Connect spec.
- The `Interceptors` field exposes connect-go's interceptor chain so callers can plug in tracing, logging, retries, or auth header injection without wrapping the HTTPClient.
- The internal per-procedure client cache uses custom codecs (`dynamicProtoCodec` / `dynamicJSONCodec`) that bind a `dynamicpb.Message` to the expected response descriptor at Unmarshal time — connect-go's default codec would otherwise produce a descriptor-less empty message.
- New `transport_connect_test.go`: covers Invoke over both wire formats, header passthrough (including binary), error mapping (`*connect.Error` is preserved through `%w` wrapping so callers can `errors.As`), context cancellation, base-URL trailing-slash handling, custom `HTTPClient`, Connect interceptors, and the `HTTPClient: nil` default-fallback path.
- New `grpc_datasource_connect_test.go`: end-to-end tests that mirror the existing `Test_DataSource_Load_WithMockService` happy path but route the call through `NewConnectTransport` against an `httptest`/H2C server backed by `MockServiceConnect`. One test per wire format. Proves the full pipeline (compiler → JSON builder → transport → response unmarshal) is transport-agnostic.

### `pkg/engine/datasource/graphql_datasource` — RPC transport factory

- `NewFactoryRPCTransport(ctx, transport)` lets the GraphQL datasource planner consume any `grpc_datasource.RPCTransport`, including a Connect transport built with `NewConnectTransport(ConnectTransportConfig{...})`.
- Existing gRPC factory entry points remain available; they wrap their gRPC clients into `RPCTransport` internally so the planner only deals with the transport abstraction.
- Datasource configuration continues to use `GRPCConfiguration` for protobuf mapping/compiler data. There is no separate Connect datasource configuration; Connect endpoint, encoding, interceptors, and HTTP client options live on `NewConnectTransport(ConnectTransportConfig{...})`.

### `pkg/engine/datasource/grpc_datasource/transport.go` — contract docs

The `RPCTransport` interface now documents the `methodFullName` format (`/package.Service/Method`, leading slash required) and the input/output expectations, so both implementations agree on the shape.

## Backward compatibility

Existing HTTP and gRPC datasource factory APIs remain compatible. For GraphQL datasource callers that want Connect, the transport is now supplied explicitly via `NewFactoryRPCTransport(ctx, grpcdatasource.NewConnectTransport(...))`; there is no separate Connect datasource configuration consumed by the planner.

## Testing

- `pkg/engine/datasource/grpc_datasource`: full test suite passes, including Connect transport coverage for both wire formats, metadata forwarding, error handling, cancellation, base URL normalization, custom `HTTPClient`, interceptors, and MockService e2e paths.
- `pkg/engine/datasource/graphql_datasource`: test suite passes with the RPC transport factory changes.
- `examples/federation`: module tests pass after dependency tidy.

## Dependencies

- Added: `connectrpc.com/connect v1.19.2`
- `golang.org/x/net` promoted from indirect to direct (the e2e test uses `http2`/`h2c` to run the Connect server over a real HTTP/2 endpoint).

## Checklist

- [x] I have discussed my proposed changes in an issue and have received approval to proceed.
- [x] I have followed the coding standards of the project.
- [x] Tests or benchmarks have been added or updated.
- [ ] Documentation has been updated on https://github.com/wundergraph/docs-website — N/A: no public-facing API change worth surfacing in docs at this stage; user-facing docs land with the consuming cosmo PR.
- [x] I have read the [Contributors Guide](https://github.com/wundergraph/cosmo/blob/main/CONTRIBUTING.md).

## Open Source AI Manifesto

This project follows the principles of the [Open Source AI Manifesto](https://human-oss.dev). Please ensure your contribution aligns with its principles.

@coderabbitai summary

### CODE REVIEW

**Comment by fengyuwusong** [comment]
Hi @Noroth , Could you please take a look? Thanks!

**Comment by Noroth** [comment]
Hi @fengyuwusong can you let me know how you planned the next steps preferably in your own words please.
I cannot be 100% confident when I don't know if the response is coming from you or Claude. I just wanna make sure we both are aligned.

This PR only adds support for the connect protocol but currently it cannot yet be wired up to the router because that part would be in the graphql_datasource.

Are you going to raise a new PR then when this was merged because you wanted to keep the changes small or was it forgotten?

I'm fine with both approaches just want to know how you planned the implementation process.

Thanks for all the work so far!

**Comment by fengyuwusong** [comment]
> Hi @fengyuwusong can you let me know how you planned the next steps preferably in your own words please. I cannot be 100% confident when I don't know if the response is coming from you or Claude. I just wanna make sure we both are aligned.
> 
> This PR only adds support for the connect protocol but currently it cannot yet be wired up to the router because that part would be in the graphql_datasource.
> 
> Are you going to raise a new PR then when this was merged because you wanted to keep the changes small or was it forgotten?
> 
> I'm fine with both approaches just want to know how you planned the implementation process.
> 
> Thanks for all the work so far!


  Hello @Noroth, I am sorry about that. Because my English is poor, so I use Claude to improve my words.

  Now I try to use my own words to explain the next steps about how to support the ConnectRPC protocol.

  - graphql-go-tools repo
    - #1509 (this PR): implement the connect protocol, and also wire it up in graphql_datasource (added based on your comment, originally planned as a separate PR).

  - cosmo repo
    - wundergraph/cosmo#2841: because ConnectRPC can use http/https protocol, so controlplane should accept http(s) URLs for GRPC_SERVICE subgraphs
    - next new PR:
      - router
        - new grpc_protocol configuration block
        - wire the new transport into factoryresolver / graph_server
        - document ConnectRPC support
        - add e2e subgraph tests
      - demo
        - write the demo project code, support both gRPC and ConnectRPC
      - examples
        - new router-grpc-connect example, showing the grpc_protocol configuration and how to point the router at a ConnectRPC subgraph

  The above is my rough plan. Please see if it matches your expectations. Thanks for your patient review. If you have any questions, feel free to ask anytime~

**Comment by fengyuwusong** [comment]
This PR is done, please feel free to take a look. Thank you @Noroth!

**Comment by Noroth** [comment]
Hi @fengyuwusong,
thanks for your explanation, and sorry for the delay. I will take a look as soon as possible.
Also your English is absolutely fine and your plan makes sense to me, thanks!

I appreciate all the work you're doing!

**Comment by fengyuwusong** [comment]
Pushed two follow-up commits to resolve the merge conflict with master and address a couple of lint findings that surfaced along the way.

Thanks for the review, @Noroth ! Happy to clarify or address any follow-up feedback.

**Comment by fengyuwusong** [comment]
hi @Noroth , we have any progress about this? Feel free talk to me if have any question, thanks~

File: v2/pkg/engine/datasource/graphql_datasource/graphql_datasource.go [hunk 1]
```diff
@@ -360,7 +361,31 @@ func (p *Planner[T]) ConfigureFetch() resolve.FetchConfiguration {
...
+		// we surface a hard error instead of silently dialing gRPC. Both
+		// transports share the same proto mapping and compiler — only the
+		// wire format differs.
+		var transport grpcdatasource.RPCTransport
+		switch {
+		case p.connectTransport != nil:
+			transport = p.connectTransport
+		case p.config.IsConnect():
+			p.stopWithError(errors.WithStack(errors.New("connect configuration requires a connect transport (use NewFactoryConnect)")))
+			return resolve.FetchConfiguration{}
```
Comment (Noroth) [inline v2/pkg/engine/datasource/graphql_datasource/graphql_datasource.go]: I'm not sure if I understand the cases correctly.
To me it looks like we need one `case` 

e.g.:

```go
case p.config.IsConnect():
    if p.connectTransport == nil {
        p.stopWithError(...)
        return ...
    }
    transport = p.connectTransport
```

But maybe I'm misunderstanding the logic here?

**Comment by Noroth** [comment]
Hey @fengyuwusong 
Sorry for the delay. I was a bit tight on time last week and I will take care of it in the upcoming days.

File: v2/pkg/engine/datasource/grpc_datasource/grpc_datasource_connect_test.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
+package grpcdatasource
```
Comment (Noroth) [inline v2/pkg/engine/datasource/grpc_datasource/grpc_datasource_connect_test.go:1]: I think in general we should have a couple more tests for edge cases.

things like 
* Inline Fragments
* Field resolvers
* Federation

You can look at the tests in `grpc_datasource_test.go` and `grpc_datasource_federation_test.go`

**Review by Noroth — CHANGES_REQUESTED** [review]
So far this looks pretty good. We should have a couple more tests for the actual execution to ensure that the implementation covers edge cases and those are also working properly.

Apart from that in parallel I'm working on this topic: https://github.com/wundergraph/graphql-go-tools/pull/1474 where I have to make sure to properly integrate your implementation then.

**Comment by fengyuwusong** [comment]
thanks for your reply, i will finish today.

File: v2/pkg/engine/datasource/graphql_datasource/graphql_datasource.go (hunk 1 above)
Comment (fengyuwusong) [reply v2/pkg/engine/datasource/graphql_datasource/graphql_datasource.go]: What I intended was to prioritize the initialized value over the configuration setting. Specifically, use `p.connectTransport` if it has been initialized.

But You're right — that single-case form is cleaner and make sense. I've pushed the change:

the planner now decides on` Configuration.IsConnect() `only, and the factory must supply the matching resource (`connectTransport` for **Connect**, `grpcClient` for **gRPC**); 

a mismatch surfaces as a hard error at plan time.

File: v2/pkg/engine/datasource/grpc_datasource/grpc_datasource_connect_test.go (hunk 2 above)
Comment (fengyuwusong) [reply v2/pkg/engine/datasource/grpc_datasource/grpc_datasource_connect_test.go:1]: I've added relevant use cases in [4689ab4](https://github.com/wundergraph/graphql-go-tools/pull/1509/commits/4689ab4ca03489c64cf1b91a272824c5cb0c0b0c0c). Thanks for pointing this out!

**Comment by fengyuwusong** [comment]
I've finished all necessary changes. Could you please take a look when you have time, @Noroth? Thanks a lot.

Also, I've reviewed the implementation of #1474. Our work can be largely integrated into it. 
Please kindly assess this. Thank you very much.

File: v2/pkg/engine/datasource/grpc_datasource/transport.go [hunk 3]
```diff
@@ -9,7 +9,16 @@ import (
...
+// Both gRPC and Connect implement this interface.
+//
+// Invoke dispatches a unary call against the remote service.
+//   - methodFullName is the gRPC-style procedure path "/package.Service/Method"
+//     (leading slash required). The gRPC transport passes it directly to
+//     grpc.ClientConnInterface.Invoke; the Connect transport appends it to
+//     the configured base URL.
+//   - input must be a *dynamicpb.Message populated by the caller.
+//   - output must be a *dynamicpb.Message bound to the expected response
+//     descriptor; Invoke populates it on success.
```
Comment (dkorittki) [inline v2/pkg/engine/datasource/grpc_datasource/transport.go]: Can you move this godoc just above the Invoke method?

File: v2/pkg/grpctest/mockservice_connect.go [hunk 4]
```diff
@@ -0,0 +1,899 @@
...
+// code generator because the adapter is a pure passthrough.
+
+package grpctest
+
+import (
+	"context"
+	"net/http"
+	"strings"
+
+	"connectrpc.com/connect"
```
Comment (dkorittki) [inline v2/pkg/grpctest/mockservice_connect.go:13]: You need to add this dependency to `examples/federation/go.(mod|sum)` in this repository as well. Head to `examples/federation` and run `go mod tidy` to add it

File: v2/pkg/engine/datasource/grpc_datasource/transport_connect_test.go [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+
+	err := transport.Invoke(ctx, "/productv1.ProductService/QueryCategory", inputMsg, outputMsg)
+	require.Error(t, err)
+	require.True(t, errors.Is(err, context.Canceled), "expected context.Canceled in error chain, got %v", err)
+}
+
+// TestConnectTransport_DefaultHTTPClient verifies that omitting HTTPClient
+// from ConnectTransportConfig falls back to http.DefaultClient and still
+// produces a working transport.
+func TestConnectTransport_DefaultHTTPClient(t *testing.T) {
```
Comment (dkorittki) [inline v2/pkg/engine/datasource/grpc_datasource/transport_connect_test.go:315]: In contrast to this test can you add a new test verifying that if you configure a custom http client that it is indeed being used?<br>Can you also add a test to verify interceptors are being used?

File: v2/pkg/engine/datasource/graphql_datasource/configuration.go [hunk 6]
```diff
@@ -29,7 +41,8 @@ type Configuration struct {
 	schemaConfiguration    SchemaConfiguration
 	customScalarTypeFields []SingleTypeField
 
-	grpc *grpcdatasource.GRPCConfiguration
+	grpc    *grpcdatasource.GRPCConfiguration
+	connect *ConnectConfiguration
```
Comment (dkorittki) [inline v2/pkg/engine/datasource/graphql_datasource/configuration.go]: Can you move this type into the `grpcdatasource` package? All of the gRPC and ConnectRPC code lives there and it reflects what `grpcdatasource.GRPCConfiguration` does as well.

File: v2/pkg/engine/datasource/graphql_datasource/graphql_datasource.go [hunk 7]
```diff
@@ -91,8 +91,9 @@ type Planner[T Configuration] struct {
 
 	minifier *astminify.Minifier
 
-	// gRPC
-	grpcClient grpc.ClientConnInterface
+	// gRPC / Connect
+	grpcClient       grpc.ClientConnInterface
+	connectTransport grpcdatasource.RPCTransport
```
Comment (dkorittki) [inline v2/pkg/engine/datasource/graphql_datasource/graphql_datasource.go]: `grpcdatasource.RPCTransport` is meant to hide both gRPC and Connect transports behind one interface but here we have two distinct fields for both. `grpcClient` keeps being used as the gRPC transport and `connectTransport` is only used for Connect transport. This results in a leaky abstraction.

I would like to replace both fields with `rpcTransport` serving both implementations.

```
type Planner[T Configuration] struct {
  // ...
	rpcTransport grpcdatasource.RPCTransport
}
```

Then you can get rid of choosing the correct implementation in `ConfigureFetch()` . Then to make it work you need to stretch the abstraction into the planner factory as well by replacing `grpcClient` and `grpcClientProvider` .

```
type Factory[T Configuration] struct {
	// ...
	//grpcClient         grpc.ClientConnInterface // delete this
	//grpcClientProvider func() grpc.ClientConnInterface // delete this
	rpcTransport         grpcdatasource.RPCTransport // new
	rpcTransportProvider func() grpcdatasource.RPCTransport // new
}
```

Then you have to adjust the planner factories constructors `NewFactoryGRPC(...)`, `NewFactoryGRPCClientProvider(...)` and `NewFactoryConnect(...)` .  Their function signatures must not change but inside those functions you need to convert the actual client to the `RPCTransport` interface:

```
func NewFactoryGRPC(executionContext context.Context, grpcClient grpc.ClientConnInterface) (*Factory[Configuration], error) {
	// ...

	return &Factory[Configuration]{
		// ...
		rpcTransport: grpcdatasource.NewGRPCTransport(grpcClient),
	}, nil
}
```

```
func NewFactoryGRPCClientProvider(executionContext context.Context, clientProvider func() grpc.ClientConnInterface) (*Factory[Configuration], error) {
	// ...

  // wrap clientProvider func in a new func
	transportProvider := func() grpcdatasource.RPCTransport {
		return grpcdatasource.NewGRPCTransport(clientProvider())
	}

	return &Factory[Configuration]{
    // ...
		connectTransportProvider: transportProvider,
	}, nil
}
```

For `NewFactoryConnect` I have a dedicated comment since we should change the signature as well.

Please let me know if you need additional help with this :-)

File: v2/pkg/engine/datasource/graphql_datasource/graphql_datasource.go [hunk 8]
```diff
@@ -1799,6 +1823,26 @@ func NewFactoryGRPCClientProvider(executionContext context.Context, clientProvid
 	}, nil
 }
 
+// NewFactoryConnect creates a Connect protocol factory for the GraphQL
+// datasource planner. The supplied RPCTransport (typically built via
+// grpcdatasource.NewConnectTransport) is dialed in place of a gRPC client
+// connection; the rest of the data source pipeline (mapping, compiler, JSON
+// builder) is identical to NewFactoryGRPC.
+func NewFactoryConnect(executionContext context.Context, connectTransport grpcdatasource.RPCTransport) (*Factory[Configuration], error) {
```
Comment (dkorittki) [inline v2/pkg/engine/datasource/graphql_datasource/graphql_datasource.go]: We should make this constructor abstract as well. When you implement my comment from above this constructor already will be abstract, just the naming needs to change:

`func NewFactoryRPCTransport(executionContext context.Context, rpcTransport grpcdatasource.RPCTransport) (*Factory[Configuration], error)`

Callers of this function then can provide `rpcTransport` with either gRPC or Connect implementation. It makes `NewFactoryGRPC(...)` redundant but we should keep it for compatibility reasons.

File: v2/pkg/engine/datasource/graphql_datasource/configuration.go [hunk 9]
```diff
@@ -46,8 +59,8 @@ func NewConfiguration(input ConfigurationInput) (Configuration, error) {
 
 	cfg.schemaConfiguration = *input.SchemaConfiguration
 
-	if input.Fetch == nil && input.Subscription == nil && input.GRPC == nil {
-		return Configuration{}, errors.New("fetch or subscription or grpc configuration is required")
+	if input.Fetch == nil && input.Subscription == nil && input.GRPC == nil && input.Connect == nil {
+		return Configuration{}, errors.New("fetch or subscription or grpc or connect configuration is required")
```
Comment (dkorittki) [inline v2/pkg/engine/datasource/graphql_datasource/configuration.go]: Error message starts to become weird. Suggestion: `fetch / subscription / grpc / connect configuration is required`

**Review by dkorittki — CHANGES_REQUESTED** [review]
Thanks for the contribution! The idea to use an abstraction for the transports is good and the code is good as well overall. There are some places which need adjustments. I added comments for these. Once that's settled we can merge it

**Comment by fengyuwusong** [comment]
Thanks again for the review. I’ve addressed the latest comments and pushed the updates.

Main changes:
- the GraphQL datasource planner/factory now uses a single `RPCTransport` abstraction;
- `NewFactoryConnect` was replaced by the more general `NewFactoryRPCTransport`;
- the separate Connect datasource config was removed, so Connect options now live on `NewConnectTransport(ConnectTransportConfig{...})`;
- added tests for custom HTTP client and interceptors;
- added the missing `examples/federation` dependency tidy changes.

Could you please take another look when you have time?

**Comment by dkorittki** [comment]
@fengyuwusong There seems to be a formatting error, as reported in the CI run https://github.com/wundergraph/graphql-go-tools/actions/runs/29098857006/job/86382426208?pr=1509#step:4:39

Can you fix this? Other than that it looks good

**Comment by fengyuwusong** [comment]
> /job/86382426208?pr=1509#step:4:39\n\nCan you fix this? Other than that

ok，I will fix it ASAP.

**Comment by fengyuwusong** [comment]
@dkorittki pls review again, thanks~

**Review by dkorittki — APPROVED** [review]

**Review by Noroth — APPROVED** [review]
